### PR TITLE
Add logging for unsubscribe links

### PR DESCRIPTION
**Pull request by VenkataNeeharikaKondipati**
This PR adds logs for unsubscribe and manage preference links. 
Adds validation for these links to check for the query params and adds logs and metrics in case some of the params are missing

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [ ] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts [hunk 1]
```diff
@@ -327,6 +326,54 @@ export class SendEmailPerformer extends MessageSendPerformer<Settings, Payload>
...
+
+    const parsedLink = new URL(link)
+    const contactId = parsedLink.searchParams.get('contactId')
+    if (!contactId || contactId == '') {
+      this.logger.error(`contactId is missing: ${link} ${workspaceId} ${audienceId}`)
+      this.statsClient.incr('missing_query_param', 1, ['param:contactId', `audienceId:${audienceId}`])
+    }
+    const data = parsedLink.searchParams.get('data')
+    if (!data || data == '') {
+      this.logger.error(`contactId is missing: ${link} ${workspaceId} ${audienceId}`)
```
Comment (Swaty-G) [inline packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:347]: should this be `data is missing:`

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts [hunk 2]
```diff
@@ -327,6 +326,54 @@ export class SendEmailPerformer extends MessageSendPerformer<Settings, Payload>
...
+    if (!wrkspaceId || wrkspaceId == '') {
+      this.logger.error(`wrkspaceId is missing: ${link} ${workspaceId} ${audienceId}`)
+      this.statsClient.incr('missing_query_param', 1, ['param:wrkspaceId', `audienceId:${audienceId}`])
+    }
+    const messageId = parsedLink.searchParams.get('messageId')
+    if (!messageId || messageId == '') {
+      this.logger.error(`messageId is missing: ${link} ${workspaceId} ${audienceId}`)
+      this.statsClient.incr('missing_query_param', 1, ['param:messageId', `audienceId:${audienceId}`])
+    }
+    const userAgent = parsedLink.searchParams.get('user-agent')
```
Comment (Swaty-G) [inline packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:370]: should this be: 
const userAgent = parsedLink.searchParams.get(`'user_agent'`) ?

**Comment by Swaty-G** [comment]
I just have the two observations above.

The below one is just a refactored version of your method to avoid repetitive checks. But the above one looks good too for logging. 

```
validateLinkAndLog(link: string): void {
  let workspaceId = this.payload.customArgs && this.payload.customArgs['workspace_id']
  let audienceId =
    this.payload.customArgs &&
    (this.payload.customArgs['audience_id'] || this.payload.customArgs['__segment_internal_audience_id__'])
  workspaceId = JSON.stringify(workspaceId)
  audienceId = JSON.stringify(audienceId)

  this.logger.info(`Validating the link: ${link} ${workspaceId} ${audienceId}`)

  const parsedLink = new URL(link)

  // Generic function to check for missing parameters
  const checkParam = (paramName) => {
    const paramValue = parsedLink.searchParams.get(paramName)
    if (!paramValue || paramValue === '') {
      this.logger.error(`${paramName} is missing: ${link} ${workspaceId} ${audienceId}`)
      this.statsClient.incr('missing_query_param', 1, [`param:${paramName}`, `audienceId:${audienceId}`])
    }
  }

  // List of required query parameters
  const requiredParams = ['contactId', 'data', 'code', 'spaceId', 'workspaceId', 'messageId', 'user_agent']

  // Check each required parameter
  requiredParams.forEach(param => checkParam(param))
}
```

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts (hunk 2 above)
Comment (VenkataNeeharikaKondipati) [reply packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:370]: It is user-agent

**Review by Swaty-G — APPROVED** [review]

**Review by Swaty-G — APPROVED** [review]

**Comment by joe-ayoub-segment** [comment]
Hi @VenkataNeeharikaKondipati can you let me know when the PR is ready for review please? and when you'd like it deployed. 

The next deploy is Tuesday next week. It's the last Partner deploy of the year.

**Comment by VenkataNeeharikaKondipati** [comment]
> Hi @VenkataNeeharikaKondipati can you let me know when the PR is ready for review please? and when you'd like it deployed.
> 
> The next deploy is Tuesday next week. It's the last Partner deploy of the year.

@joe-ayoub-segment we want this to be deployed tomorrow

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts [hunk 3]
```diff
@@ -327,6 +327,36 @@ export class SendEmailPerformer extends MessageSendPerformer<Settings, Payload>
...
+
+    this.logger.info(`Validating the link: ${link} ${workspaceId} ${audienceId}`)
+
+    const parsedLink = new URL(link)
+    // Generic function to check for missing parameters
+    const checkParam = (paramName: string) => {
+      const paramValue = parsedLink.searchParams.get(paramName)
+      if (!paramValue || paramValue === '') {
+        console.log('entered')
+        console.log(audienceId)
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:347]: Hi there! Any specific reason why we are mixing calls to `console.log` and `this.logger` ? they both output to the logs so we should aim to stick to one. You can change the level using any of [these](https://github.com/segmentio/action-destinations/blob/418550fb74af1c8117995e49b87bf90803b9133a/packages/core/src/destination-kit/index.ts#L355) methods.

Comment (marinhero) [reply packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:347]: Also, we would really appreciate an updated description in the PR 🙏🏽

**Review by vikramkumar19 — APPROVED** [review]
LGTM! agreed with @marinhero on utilizing same method for logger

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts (hunk 3 above)
Comment (VenkataNeeharikaKondipati) [reply packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts]: Good catch, console logs was for local testing. Removed them and updated the decsriptioni

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts [hunk 4]
```diff
@@ -327,6 +327,34 @@ export class SendEmailPerformer extends MessageSendPerformer<Settings, Payload>
     }, {})
   }
 
+  @track()
+  validateLinkAndLog(link: string): void {
+    let workspaceId = this.payload.customArgs && this.payload.customArgs['workspace_id']
+    let audienceId =
+      this.payload.customArgs &&
+      (this.payload.customArgs['audience_id'] || this.payload.customArgs['__segment_internal_audience_id__'])
+    workspaceId = JSON.stringify(workspaceId)
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:336]: Nit: these are strings aren't they? if so, you don't need to `stringify` them.

File: packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts [hunk 5]
```diff
@@ -327,6 +327,34 @@ export class SendEmailPerformer extends MessageSendPerformer<Settings, Payload>
...
+
+    this.logger.info(`Validating the link: ${link} ${workspaceId} ${audienceId}`)
+
+    const parsedLink = new URL(link)
+    // Generic function to check for missing parameters
+    const checkParam = (paramName: string) => {
+      const paramValue = parsedLink.searchParams.get(paramName)
+      if (!paramValue || paramValue === '') {
+        this.logger.error(`${paramName} is missing: ${link} ${workspaceId} ${audienceId}`)
+        this.statsClient.incr('missing_query_param', 1, [`param:${paramName}`, `audienceId:${audienceId}`])
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/engage/sendgrid/sendEmail/SendEmailPerformer.ts:347]: Nice! passing these as tags will prevent creating a multitude of metrics. Just a small nit, consider rewording `param` and `audienceId` to something more meaningful to your use case.

**Review by marinhero — APPROVED** [review]
Thank you for addressing my comments. I left a couple of nits. Up to you to address them or not. Nice work!

**Comment by joe-ayoub-segment** [comment]
Hi @VenkataNeeharikaKondipati this PR was deployed. No Push was done

**Pull request merged**